src/JukeBox.cpp

```diff
@@ -30,49 +49,294 @@ bool JukeBox::start(string exeName, string configFilePath) {
+    return Error(ErrorCode::InvalidFormat, "Track found in both queues.");
+  }
+
+  QueueType q;
+  if (trackFoundAdmin)
+    q = QueueType::Admin;
+  if (trackFoundNormal)
+    q = QueueType::Normal;
+
+  auto retTrack = mDataStore->removeTrack(trkid, q);
+  if (holds_alternative<Error>(retTrack))
+    return get<Error>(retTrack);
+
+  return nullopt;
 }
 
 TResultOpt JukeBox::moveTrack(TSessionID const &sid,
                               TTrackID const &trkid,
-                              QueueType type) {
-  return Error(ErrorCode::NotImplemented, "moveTrack is not implemented yet");
+                              QueueType toQueue) {
+  auto retUser = mDataStore->getUser(sid);
+  if (holds_alternative<Error>(retUser))
+    return get<Error>(retUser);
+  User user = get<User>(retUser);
+
+  if (!user.isAdmin) {
+    LOG(WARNING) << "JukeBox.moveTrack: User with session ID '" << sid
+                 << "' and nickname '" << user.Name
+                 << "' is not priviledged to move a track.";
+    return Error(ErrorCode::AccessDenied, "User is not an admin.");
+  }
+
+  /* Moving a track to another queue means deleting it
+   * in the respective other */
+  QueueType fromQueue;
+  if (toQueue == QueueType::Admin)
+    fromQueue = QueueType::Normal;
+  if (toQueue == QueueType::Normal)
+    fromQueue = QueueType::Admin;
+
+  /* Query the source queue for the track that is to be deleted */
+  auto retQueue = mDataStore->getQueue(fromQueue);
+  if (holds_alternative<Error>(retQueue))
+    return get<Error>(retQueue);
+  Queue queue = get<Queue>(retQueue);
+  bool trackFound = get<bool>(mDataStore->hasTrack(trkid, fromQueue));
+
+  if (!trackFound) {
+    LOG(WARNING) << "Jukebox.moveTrack: TrackID '" << trkid
+                 << "' could not be found.";
+    return Error(ErrorCode::DoesntExist, "Track not found.");
+  }
+
+  /* Remove from source queue */
+  auto retTrack = mDataStore->removeTrack(trkid, fromQueue);
+  if (holds_alternative<Error>(retTrack))
+    return get<Error>(retTrack);
+
+  /* Add to target queue */
+  auto res = mDataStore->addTrack(get<BaseTrack>(retTrack), toQueue);
+  if (res.has_value())
+    return res.value();
+
+  return nullopt;
 }
 
 TResultOpt JukeBox::controlPlayer(TSessionID const &sid, PlayerAction action) {
-  return Error(ErrorCode::NotImplemented,
-               "controlPlayer is not implemented yet");
+  int const volChangePercent = 10;
+  auto retUser = mDataStore->getUser(sid);
+  if (holds_alternative<Error>(retUser))
+    return get<Error>(retUser);
+  User user = get<User>(retUser);
+
+  if (!user.isAdmin) {
+    LOG(WARNING) << "JukeBox.controlPlayer: User with session ID '" << sid
+                 << "' and nickname '" << user.Name
+                 << "' is not priviledged to control the player.";
+    return Error(ErrorCode::AccessDenied, "User is not an admin.");
+  }
+
+  TResultOpt ret = nullopt;
+  TResult<size_t> volume;
+  TResult<QueuedTrack> playingTrk;
+
+  switch (action) {
```

Comment:
in this switch you drop some return values (`ret`) without handling them (see `Skip`, `VolumeUp`, `VolumeDown`)